consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -326,37 +351,40 @@ func (vet *ValidatorEnodeDB) RefreshValPeers(valset istanbul.ValidatorSet, ourAd
 
 	if valset.ContainsByAddress(ourAddress) {
 		// transform address to enodeURLs
-		newEnodeURLs := []string{}
+		newNodes := []*enode.Node{}
 		for _, val := range valset.List() {
-			entry, err := vet.getAddressEntry(val.Address())
+			entry, err := vet.GetAddressEntry(val.Address())
 			if err == nil {
-				newEnodeURLs = append(newEnodeURLs, entry.enodeURL)
+				newNodes = append(newNodes, entry.Node)
 			} else if err != leveldb.ErrNotFound {
 				vet.logger.Error("Error reading valEnodeTable: GetEnodeURLFromAddress", "err", err)
 			}
 		}
 
-		vet.handler.ReplaceValidatorPeers(newEnodeURLs)
+		vet.logger.Info("hander output", "handler", vet.handler)
+		vet.handler.ReplaceValidatorPeers(newNodes)
 	} else {
 		// Disconnect all validator peers if this node is not in the valset
 		vet.handler.ClearValidatorPeers()
 	}
 }
 
 func (vet *ValidatorEnodeDB) addDeleteToBatch(batch *leveldb.Batch, address common.Address) error {
-	entry, err := vet.getAddressEntry(address)
+	entry, err := vet.GetAddressEntry(address)
 	if err != nil {
 		return err
 	}
 
 	batch.Delete(addressKey(address))
-	batch.Delete(enodeURLKey(entry.enodeURL))
-	vet.handler.RemoveValidatorPeer(entry.enodeURL)
+	batch.Delete(nodeIDKey(entry.Node.ID()))
+	if vet.handler != nil {
+		vet.handler.RemoveValidatorPeer(entry.Node)
+	}
 	return nil
 }
 
-func (vet *ValidatorEnodeDB) getAddressEntry(address common.Address) (*addressEntry, error) {
-	var entry addressEntry
+func (vet *ValidatorEnodeDB) GetAddressEntry(address common.Address) (*AddressEntry, error) {
```

Comment:
public but no readlock.
i would probably not expose this function is possbile. Why would a client need to see the view? it should see the address or node

Reply:
As we discussed, I make `GetAddressEntry` private (e.g. I renamed to `getAddressEntry`).